File: Client/Classes/CombatDSL/Interpreter/Parser.cs
using System.Collections.Generic;


public sealed class Parser
{
    private readonly IReadOnlyList<Token> _tokens;
    private int _position;

    private Parser(IReadOnlyList<Token> tokens)
    {
        _tokens = tokens;
    }

    public static Expr Parse(string text)
    {
        Lexer lexer = new(text);
        IReadOnlyList<Token> tokens = lexer.Lex();

        Parser parser = new(tokens);
        Expr expr = parser.ParseExpr();

        parser.Expect(TokenKind.End, "Expected end of input.");

        return expr;
    }

    private Expr ParseExpr()
    {
        Token token = Current;

        return token.Kind switch
        {
            TokenKind.Integer => ParseInt(),
            TokenKind.Identifier => ParseCall(),
            _ => throw new DslParseException("Expected integer literal or function call.", token.Position)
        };
    }

    private Expr ParseInt()
    {
        Token token = Expect(TokenKind.Integer, "Expected integer literal.");
        return new IntLit(token.IntValue, new TextSpan(token.Position, token.Text.Length));
    }

    private Expr ParseCall()
    {
        Token name = Expect(TokenKind.Identifier, "Expected function name.");
        Expect(TokenKind.LeftParen, $"Expected '(' after '{name.Text}'.");

        List<Expr> args = new();

        if (!Match(TokenKind.RightParen))
        {
            while (true)
            {
                args.Add(ParseExpr());

                if (!Match(TokenKind.Comma))
                    break;
            }

            Expect(TokenKind.RightParen, $"Expected ')' after arguments for '{name.Text}'.");
        }

        int end = Previous.Position + Previous.Text.Length;
        return new Call(
            name.Text,
            args,
            new TextSpan(name.Position, end - name.Position)
        );
    }

    private Token Current => _tokens[_position];

    private Token Previous => _tokens[_position - 1];

    private bool Match(TokenKind kind)
    {
        if (Current.Kind != kind)
            return false;

        _position++;
        return true;
    }

    private Token Expect(TokenKind kind, string message)
    {
        if (Current.Kind != kind)
            throw new DslParseException(message, Current.Position);

        Token token = Current;
        _position++;
        return token;
    }
}

File: server/internal/server/combatDSL/Parser.go
package combatDSL

import tokens2 "server/internal/server/combatDSL/coreAST/tokens"

type Parser struct {
	tokens   []tokens2.Token
	position int
}

func Parse(text string) (tokens2.Expr, error) {
	lexer := NewLexer(text)

	tokens, err := lexer.Lex()
	if err != nil {
		return nil, err
	}

	parser := &Parser{tokens: tokens}

	expr, err := parser.parseExpr()
	if err != nil {
		return nil, err
	}

	if _, err := parser.expect(tokens2.TokenEnd, "expected end of input"); err != nil {
		return nil, err
	}

	return expr, nil
}

func (p *Parser) parseExpr() (tokens2.Expr, error) {
	token := p.current()

	switch token.Kind {
	case tokens2.TokenInteger:
		return p.parseInt()
	case tokens2.TokenIdentifier:
		return p.parseCall()
	default:
		return nil, &tokens2.ParseError{
			Message:  "expected integer literal or function call",
			Position: token.Position,
		}
	}
}

func (p *Parser) parseInt() (tokens2.Expr, error) {
	token, err := p.expect(tokens2.TokenInteger, "expected integer literal")
	if err != nil {
		return nil, err
	}

	return tokens2.NewIntLit(
		token.IntValue,
		tokens2.TextSpan{Start: token.Position, Length: len(token.Text)},
	), nil
}

func (p *Parser) parseCall() (tokens2.Expr, error) {
	name, err := p.expect(tokens2.TokenIdentifier, "expected function name")
	if err != nil {
		return nil, err
	}

	if _, err := p.expect(tokens2.TokenLeftParen, "expected '(' after '"+name.Text+"'"); err != nil {
		return nil, err
	}

	args := make([]tokens2.Expr, 0)

	if !p.match(tokens2.TokenRightParen) {
		for {
			arg, err := p.parseExpr()
			if err != nil {
				return nil, err
			}

			args = append(args, arg)

			if !p.match(tokens2.TokenComma) {
				break
			}
		}

		if _, err := p.expect(tokens2.TokenRightParen, "expected ')' after arguments for '"+name.Text+"'"); err != nil {
			return nil, err
		}
	}

	previous := p.previous()
	end := previous.Position + len(previous.Text)

	return tokens2.NewCall(
		name.Text,
		args,
		tokens2.TextSpan{Start: name.Position, Length: end - name.Position},
	), nil
}

func (p *Parser) current() tokens2.Token {
	return p.tokens[p.position]
}

func (p *Parser) previous() tokens2.Token {
	return p.tokens[p.position-1]
}

func (p *Parser) match(kind tokens2.TokenKind) bool {
	if p.current().Kind != kind {
		return false
	}

	p.position++
	return true
}

func (p *Parser) expect(kind tokens2.TokenKind, message string) (tokens2.Token, error) {
	if p.current().Kind != kind {
		return tokens2.Token{}, &tokens2.ParseError{
			Message:  message,
			Position: p.current().Position,
		}
	}

	token := p.current()
	p.position++

	return token, nil
}
